### ISSUE TITLE
utils script to standardise evaluation

### ISSUE THREAD

**Issue by dylannn02** [joelchunsheng/50.039-Deep-Learning#10]
## Motivation

Each notebook currently duplicates the same evaluation boilerplate — metric collection loops, curve plots, threshold tuning, confusion matrix display — with minor inconsistencies between them (e.g. some track recall/F2, some don't; accuracy is used despite the dataset being class-imbalanced). A shared utility module eliminates this duplication and ensures all experiments are evaluated consistently.

Additionally, **raw accuracy is a misleading metric** for the HAM10000 dataset due to class imbalance (melanoma is the minority class). Balanced accuracy (average recall per class) is a more appropriate measure and will replace accuracy across all tracked metrics.

### New: `src/utils/evaluation.py`

Four functions notebooks can import directly:

- **`plot_training_curves(train_history, val_history)`** — plots a 2×2 grid (Loss, Balanced Accuracy, Recall, F2) from the per-epoch metric dicts returned by the training loop. Replaces 4+ separate `plt.plot(...)` blocks per notebook.

- **`get_predictions(model, dataloader, device, threshold=0.5)`** — runs the model on a dataloader and returns `(labels, probs, preds)`. Reusable for both threshold tuning and final evaluation.

- **`find_best_threshold(model, val_loader, device)`** — sweeps thresholds 0.01–0.90 to find the value that maximises F2 score on the validation set. Returns `(best_threshold, best_f2)`.

- **`evaluate_model(model, test_loader, device, threshold=0.5)`** — full test evaluation: prints classification report + AUC-ROC, and plots a labelled confusion matrix (`["Non-Melanoma", "Melanoma"]`).

### New: `src/utils/__init__.py`

Exports the above four functions for clean imports.

### Updated: `src/training/trainer.py`

`train_one_epoch` and `validate_one_epoch` will return `balanced_accuracy` instead of `accuracy`, keeping training-time metrics consistent with the evaluation utility.

## Intended notebook usage

```python
import sys; sys.path.append('../..')
from src.utils import plot_training_curves, find_best_threshold, evaluate_model

# After training loop
plot_training_curves(train_history, val_history)

# Threshold tuning on validation set
best_thresh, best_f2 = find_best_threshold(model, val_loader, device)

# Final test evaluation
evaluate_model(model, test_loader, device, threshold=best_thresh)
```
